A16-binary-logging.md

```diff
@@ -0,0 +1,386 @@
+representation.
+
+The proto of GrpcLogEntry:
+
+```proto
+syntax = "proto3";
+
+package grpc.binarylog.v1;
+
+import "google/protobuf/duration.proto";
+import "google/protobuf/timestamp.proto";
+
+option java_multiple_files = true;
+option java_package = "io.grpc.binarylog.v1";
+option java_outer_classname = "BinaryLogProto";
+
+// Log entry we store in binary logs
+message GrpcLogEntry {
+  // Enumerates the type of event
+  // Note the terminology is different from the RPC semantics
+  // definition, but the same meaning is expressed here.
+  enum EventType {
+    EVENT_TYPE_UNKNOWN = 0;
+    // Header sent from client to server
+    EVENT_TYPE_CLIENT_HEADER = 1;
+    // Header sent from server to client
+    EVENT_TYPE_SERVER_HEADER = 2;
+    // Message sent from client to server
+    EVENT_TYPE_CLIENT_MESSAGE = 3;
+    // message sent from server to client
+    EVENT_TYPE_SERVER_MESSAGE = 4;
+    // Signal that client is done sending
+    EVENT_TYPE_CLIENT_HALF_CLOSE = 5;
+    // Trailers sent from server to client.
+    // This marks the end of the RPC. Events may arrive after
+    // this due to races. The extra events are not considered
+    // a part of the RPC and may be safely ignored.
+    EVENT_TYPE_SERVER_TRAILER = 6;
+    // Signal that the RPC is cancelled.
+    // This marks the end of the RPC. Events may arrive after
+    // this due to races. The extra events are not considered
+    // a part of the RPC and may be safely ignored.
+    EVENT_TYPE_CANCEL = 7;
+  }
+
+  // Enumerates the entity that generates the log entry
+  enum Logger {
+    LOGGER_UNKNOWN = 0;
+    LOGGER_CLIENT = 1;
+    LOGGER_SERVER = 2;
+  }
+
+  // The timestamp of the binary log message
+  google.protobuf.Timestamp timestamp = 1;
+
+  // Uniquely identifies a call. The value must not be 0 in order to disambiguate
+  // from an unset value.
+  // Each call may have several log entries, they will all have the same call_id.
+  // Nothing is guaranteed about their value other than they are unique across
+  // different RPCs in the same gRPC process.
+  uint64 call_id = 2;
+
+  // The entry sequence id for this call. The first GrpcLogEntry has a
+  // value of 1, to disambiguate from an unset value. The purpose of
+  // this field is to detect missing entries in environments where
+  // durability or ordering is not guaranteed.
+  uint64 sequence_id_within_call = 3;
+
+  EventType type = 4;
+  Logger logger = 5;  // One of the above Logger enum
+
+  // The logger uses one of the following fields to record the payload,
+  // according to the type of the log entry.
+  oneof payload {
+    ClientHeader client_header = 6;
+    ServerHeader server_header = 7;
+    // Used by EVENT_TYPE_CLIENT_MESSAGE, EVENT_TYPE_SERVER_MESSAGE
+    Message message = 8;
+    Trailer trailer = 9;
+  }
+
+  // true if payload does not represent the full message or metadata.
+  bool payload_truncated = 10;
+
+  // Peer address information, will only be recorded in
```

Comment:
Should we put Peer into Client/ServerHeader since they are only set with those event types?

Then the comments *might* be simpler to explain: `ClientHeader.Peer: // only set by server`.  (and vice-versa)

Should we put it in the trailer, too, in case the RPC is trailers-only?

Reply:
Having the peer in one of three places might make it harder for users to understand. The trailers-only case in particular may be something people overlook when filtering the records.

I tried to rephrase the comment. Does this make it easier to understand?

Reply:
Sounds good thanks.